File: benchmarks/OCaml/relational/IsaPlanner/goal52.ml
type list = Nil | Cons of int * list

let rec count n l =
  match l with
  | Nil -> 0
  | Cons(x, xs) ->
    if n = x then 1 + count n xs
    else count n xs

let rec append l ys =
  match l with
  | Nil -> ys
  | Cons(x, xs) ->
    Cons(x, append xs ys)

let rec rev l =
  match l with
  | Nil -> Nil
  | Cons(x, xs) -> append (rev xs) (Cons(x, Nil))

let main n l =
  assert(count n l = count n (rev l))

[@@@types
"count :: (t1: int) -> (t2: list) -> {t3: int | count[0:2](t1, t2, t3) && t3 >= 0}
"]

File: benchmarks/CHC/ADTs/ADTRem-benchmarks/CHC-source-prolog/isaplanner/goal52.pl
:- pred ff.
:- pred count(int,list(int),int).
:- pred rev(list(int),list(int)).
:- pred leq(int,int,bool).
:- pred append(list(int),list(int),list(int)).

:- mode count(in,in,out).
:- mode rev(in,out).
:- mode leq(in,in,out).
:- mode append(in,in,out).

append([],A,A).
append([A|B],C,[A|D]) :- append(B,C,D).

rev([],[]).
rev([A|B],C) :- rev(B,D), append(D,[A],C).
count(A,[],B) :- B=0.
count(A,[A|B],C) :- C=1+D, count(A,B,D).
count(A,[B|C],D) :- B-A>=1, count(A,C,D).
count(A,[B|C],D) :- B-A=< -1, count(A,C,D).
ff :- A-B>=1, count(C,D,B), rev(D,E), count(C,E,A).
ff :- A-B=< -1, count(C,D,B), rev(D,E), count(C,E,A).
